octopoes/tests/conftest.py

```diff
@@ -255,3 +258,54 @@ def mock_xtdb_session():
 @pytest.fixture
 def origin_repository(mock_xtdb_session):
     yield XTDBOriginRepository(Mock(spec=EventManager), mock_xtdb_session, XTDBType.XTDB_MULTINODE)
+
+
+def seed_system(octopoes_api_connector: OctopoesAPIConnector, valid_time):
```

Comment:
And if this indeed fits better as a fixture, perhaps also a nice teardown part that removes the created observations?

Reply:
No need to teardown, the `xtdb_session` fixture takes care of that. And: it is a function now so I could conveniently run
```
docker compose exec octopoes_api python -c "from tests.conftest import *; seed_system(OctopoesAPIConnector('http://octopoes_api:80', 'q'), datetime.now(tz=timezone.utc))"
```
to have that same system data in my dev environment (for organization "q").

But to be honest I am still not really sure about using fixtures for seeding data, versus creating test doubles or state setup. For me it feels like the latter two make sense as a singleton to be re-used as fixtures, where each test usually wants (or should want) more intricate control over what data is present to perform the tests. For instance, parametrize the `seed_system` call or change one object without affecting a bunch of other tests depending on it.

Reply:
> Good point and I do appreciate the flexibility this gives. But I generally still think that using fixtures for seeding data provides a standardized and consistent approach within the pytest environment. Fixtures allow parameterization, enabling you to customize seed data for different test scenarios. This ensures that each test maintains its independence while still benefiting from the convenience of pre-defined data. Also, the reusability of fixtures across multiple tests generally results in cleaner and a more maintainable testing suite.
> > For instance, parametrize the `seed_system` call or change/remove/ad objects to the seed without affecting a bunch of other tests depending on it.
> 
> In that case, you could still use different fixtures of course. And if there's overlap in data, then that can be generalized by nesting fixtures within fixtures (we do this quite often in the Rocky tests)

I actually think having fixtures for both singletons and data seeds gets quite confusing when you have to combine data fixtures, especially if this means we should do all seeding in the fixtures. This in particular I have noticed myself working with the Rocky tests combining multiple users, organizations and groups. I have found fixtures to be nice if you don't have to think about them, re-use them a lot and need the teardown or dependency injection capabilities. If however I read a test that says `assert page.contains("testuser")`, I don't want to have to go to the `user` fixture to find out whether this is value refers to his `name`, `username` or `email`.

Fixtures do now allow parameterization in the test itself, so you could end up creating vague names to group data seeds and still not be able to say "in this test I want the user to have 'abc' as a name" without doing model updates in your test - then why seed and correct instead of seed yourself in the test?

Compare for instance the fixture-per-entity Rocky approach to `test_meta_repository.py::test_save_boefje_meta` in Bytes. In the first 11 lines we create three entities of the same type where we use the `id`, `boefje_id` and `input_ooi` arguments. This means that for this test we need at least 3 `boefje_meta` fixtures about 8 lines each:
```python3
@pytest.fixture
def boefje_meta(meta_repository: SQLMetaDataRepository) -> BoefjeMeta:
    boefje_meta = get_boefje_meta()

    with meta_repository:
        meta_repository.save_boefje_meta(boefje_meta)    

    return boefje_meta

...
```

Moreover, every combination of these arguments would require new fixtures, especially when combining it with different models. This sounds to me like a scenario where you end up with many data fixtures within data fixtures called "second_boefje_meta_with_input_ooi_and_second_test_boefje". Or just accept that there are fixtures that give you more or everything. For both scenario's I think this makes the tests actually less readable and maintainable.

Reply:
See https://github.com/minvws/nl-kat-coordination/issues/2049 :grin: